New outcome (curse) : Everyone Loves Larpa

inks007 opened the pull request:
- Chaos larpa, orbit larpa, larpa explosion and larpa bounce enabled

Differences from the mod
- Larpa bounce fixed
- Projectiles originating from the environment are not larpa'ed (e.g. from destruction of the forcefield generator in hiisi base)
- Projectiles with zero velocity are not larpa'ed (e.g. path of dark flame), with behaviour more consistent with the spell modifiers

 Changes to be committed:
	new file:   files/entities/misc/effect_everyone_loves_larpa.xml
	new file:   files/scripts/effect_everyone_loves_larpa.lua
	new file:   files/ui_gfx/status_indicators/everyone_loves_larpa_ui.png
	new file:   twitch_fragments/outcomes/everyone_loves_larpa.lua

inks007 committed:
New outcome (curse): Everyone Loves Larpa

- Chaos larpa, orbit larpa, larpa explosion and larpa bounce enabled

Differences from the mod
- Larpa bounce fixed
- Projectiles originating from the environment are not larpa'ed (e.g. from destruction of the forcefield generator in hiisi base)

 Changes to be committed:
	new file:   files/entities/misc/effect_everyone_loves_larpa.xml
	new file:   files/scripts/effect_everyone_loves_larpa.lua
	new file:   files/ui_gfx/status_indicators/everyone_loves_larpa_ui.png
	new file:   twitch_fragments/outcomes/everyone_loves_larpa.lua

inks007 committed:
Do not larpa projectiles with zero velocity
- Fixes path of dark flame spam

inks007 committed:
Do not larpa projectiles that spawn other projectiles
- Removes:
  -- Spiraalikalma/Kiukkukalma's orb spawner shots (spawned orbs will still larpa)
  -- Ylialkemisti's wand orbs (which spawns wand beams)
  -- Ylialkemisti's wand beams (spawned projectiles will still larpa, and may be excessive)
  -- Toukka's slime trail (spawned slimeblobs will still larpa)

inks007 committed:
Update list of blacklisted projectiles (all projectiles that spawn other projectiles, or spells that fragment)
Remove zero velocity fix (issues are fixed by blacklisting dark flame stationary & thunderball lightning burst fragments)

- Removed projectile spawners:
  -- Suur-Ukko's thunderball trail
  -- Jättilaser-lennokki (toaster) megalaser
  -- Kolmisilmän silmä (mecha kolmi) rocket roll
  -- Mestarien mestari (MoM) status burst
  -- Kolmisilmä fire pillar/orb homing/melee shots

- Removed projectile parts (fragments):
  -- Path of dark flame stationary trail
  -- Glitter bomb shrapnel
  -- Spiral shot parts
  -- Thunderball lightning burst

inks007 committed:
Add Huhtasieni (giga fungus) pollen ball projectile to the exclusion list (pollen ball spawns pollen)

Miczu commented:
I didn't see PR notifications... feel free to DM me on discord if you ever create a pr, so I react sooner. 
This will require a lot more testing, as it could be both instant death in many situations and quite laggy depending on situation.

Miczu commented:
Could you update manifest.json with sha256 of added/modified files? I use this site: https://www.strerr.com/en/sha256_file.html

This is probably too deadly and I don't see mentions of fixing glimmer bomb enemy lagging the game. I don't mind testing it a bit more with manifest updated, but for balance reason this could be unlikely add in current form.

If this was only one of the larpa effects at a time (random?), then it wouldn't cause such a spike in player difficult.

inks007 committed:
Randomly choose only one larpa per vote. Larpa effect types will stack with multiple votes.

 Changes to be committed:
	modified:   files/entities/misc/effect_everyone_loves_larpa.xml
	modified:   files/scripts/effect_everyone_loves_larpa.lua
	new file:   files/scripts/effect_everyone_loves_larpa_end.lua
	modified:   manifest.json
	modified:   twitch_fragments/outcomes/everyone_loves_larpa.lua

inks007 commented:
Hi Miczu,

One larpa effect at a time is a good idea.
The majority of the lags occur due to exponential larpa.

Examples of larpa being applied to a projectile, then being applied to the projectile parts:
- darkflame_stationary
- glitter_bomb_shrapnel
- spiral_part
- thunderburst_thundermage

All the other exclusions are examples of projectiles that spawn multiple other projectiles.
If needed, I could also implement a projectile limit (e.g. 250) which would turn off additional larpa effects.

Miczu commented on manifest.json:
```diff
@@ -309,5 +309,10 @@
+        "files/scripts/effect_everyone_loves_larpa.lua": "9c2b813576603c5aef563417f7de54ab4eaef57242d9244981159efd41532416",
+        "files/scripts/effect_everyone_loves_larpa_end.lua": "e105887a9f97c77252af638361b811a59ed24d27f77cb127637935ed9773ff3c",
+        "files/ui_gfx/status_indicators/everyone_loves_larpa_ui.png": "5455d320ac7667ffe4209a3da64230bea0fbfb8fb16dc03adb79b8b701b89d2c",
+        "twitch_fragments/outcomes/everyone_loves_larpa.lua": "363a18c3a23412b5047b8e5431f3e41737d9162960fd532224d725b5fdf1152d"
```
Could you move those around to maintain alphabetic order in the file? Beside esthetic code (my main reason) it could potentially help with merge conflicts - multiple PR could potentially try to add in the same spot (after last row and before brackets), so manual intervention could be needed.

Miczu replied:
Looks like there is an issue with manifest file:
`[update] Self-update started (Branch: everyone_loves_larpa)
[update] Update check started (Server: 0)
[update] Update check failed (Server: 0): FetchError: invalid json response body at https://raw.githubusercontent.com/inks007/noita-twitch-integration/everyone_loves_larpa/manifest.json reason: Unexpected string in JSON at position 36651
[update] Update check failed
[update] Self-update failed
[update] ERROR: Unable to perform self-update!`

Miczu committed:
Update manifest.json

Missing comma. Still need to be reordered, but I also want to just test it for now.

Miczu commented:
Slow larpa trail, chaos larpa is unbalanced, maybe remove altogether.

Larpa on bounce and on death is probably my favorite.
Larpa orbit is ok as is.
Larpa in random direction interactions with bombs, tnt thrower is too intense.
Larpa up/down is a bit to frequent.

I worry that stacking those larpa effects can make it quite unplayable. Especially if someone goes with really silly timers like vote each [single digit] seconds. Probably it is better to reduce effects that create more projectile by just being in flight as those feel extra bad with long lasting projectiles.

Other possible protection could be to extend the duration instead of stacking active effects. That puts a cap on how bad it can be for player and cpu.

If it was just bounce, death, orbit - it could be added with those.
Having others requires a lot more balancing out.

inks007 reviewed (commented):
Whoops, sorry. I've written bash script for the manifest.json. Should be correct for the next commit.

Larpa on bounce allows shots to ricochet around corners - quite dangerous. Larpa explosion releases those extra 8 projectiles, causing up to 9x damage - also quite dangerous. Yeah, they're the most fun.
Larpa orbit is probably the safest.
Chaos larpa is unpredictable and annoying, yes.
Copy trail really doesn't do anything... until you are shooting explosives.
Up and down larpa though, is very predictable.

The initial implementation merely overlapped the duration of the single larpa initially chosen. I really like the larpa stacking though... it's like a secret interaction if you double vote. Perhaps the way to mitigate some of this danger is to make the outcome rarer (maybe 10-20/2000). That way, the chance of getting a second larpa immediately after being voted in would be 0.5-1%. Or 1-2% within two votes. Additionally, keeping the more predictable, less lethal larpas (up/down/copy/orbit) would also serve to reduce the chance of larpa bounce/explosion/chaos being activated.

I have two more additional fixes to push with the next commit
- I think disabling the larpa effects when the projectile # > 300 is reasonable
- Fix for the high alchemist's shield causing infinite larpa

Please weigh in one more time, and I'll make whatever changes you request. I'm getting used to github now, and it'll be easy to rebalance.

inks007 committed:
Merge branch 'Miczu:master' into everyone_loves_larpa

inks007 committed:
Implemented changes as suggested by Miczu, plus performance fixes.

Enabled larpas - bounce, orbit, explosion.
Improve performance of stacked larpa outcomes.
Set larpa limit to 300.
Patch infinite larpa instance with Ylialkemisti's counter shield.

inks007 committed:
Enable OCD.

inks007 committed:
Whoops manifest json.
